### PR TITLE
Given all 4 exercises of week3.

### PR DESCRIPTION
**Pull request by shbh29**
@imranmd - PTAL

1. Implementation of Stack with closure.
2. call, bind, apply use case explained with program.
3. Output of program explained. 
4. memoize implementation for sum method.

### CODE REVIEW

File: Week-3/Ex-1/index.js [hunk 1]
```diff
@@ -0,0 +1,30 @@
+
+function memoize(fn) {
+    let cache = new Map();
+    return function(...args) {
+        const key = args.toString();
+        console.log(`key formed is: ${key}`);
```
Comment (imranmd) [inline Week-3/Ex-1/index.js:6]: Too many consoles, when raising PR try to remove the consoles which are not being used

File: Week-3/Ex-1/index.js [hunk 2]
```diff
@@ -0,0 +1,30 @@
+
+function memoize(fn) {
+    let cache = new Map();
+    return function(...args) {
+        const key = args.toString();
```
Comment (imranmd) [inline Week-3/Ex-1/index.js:5]: What about the edge cases?

What if the args is undefined or null?

File: Week-3/Ex-2/callBindApply.js [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+let methodReferenceWithBind = user.print.bind(user);
+methodReferenceWithBind();
+
+//limitation with bind is: method Reference cannot be updated for different this context.
+let bindReferenceChange = methodReferenceWithBind.bind({fname: "Imran", designation: "Mentor"});
+bindReferenceChange();
+
+// call comes to resuce. With Call you can change the this context.
+// Bind hard codes this context and creates another function.
+// Whereas, call will call same function with different this context.
```
Comment (imranmd) [inline Week-3/Ex-2/callBindApply.js:26]: Amazing understanding!

File: Week-3/Ex-4/stack.js [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+console.log("removed: ",stack.pop());
+stack.print();
+
+console.log("removed: ",stack.pop());
+stack.print();
+
+console.log("removed: ",stack.pop());
+stack.print();
+
+// items not accessible outside stack object
```
Comment (imranmd) [inline Week-3/Ex-4/stack.js:36]: Good

**Review by imranmd — APPROVED** [review]

File: Week-3/Ex-1/index.js (hunk 2 above)
Comment (shbh29) [reply Week-3/Ex-1/index.js:5]: Noted.

File: Week-3/Ex-1/index.js (hunk 1 above)
Comment (shbh29) [reply Week-3/Ex-1/index.js:6]: Noted.

File: Week-3/Ex-2/callBindApply.js (hunk 3 above)
Comment (shbh29) [reply Week-3/Ex-2/callBindApply.js:26]: Thanks

File: Week-3/Ex-4/stack.js (hunk 4 above)
Comment (shbh29) [reply Week-3/Ex-4/stack.js:36]: Thanks

**Pull request merged**